wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+-- correctly reschedule all the transaction.
+purgeSlot :: Slot -> Scheduler -> Scheduler
+purgeSlot slot (Scheduler s) = Scheduler (IntMap.delete slot s)
+
+-- TODO(adn): Better placed into module 'Cardano.Wallet.Kernel.DB.Spec'?
+
+-- | Computes the union between two 'Pending' sets.
+unionPending :: Pending -> Pending -> Pending
+unionPending (Pending new) (Pending old) =
+    Pending (M.union <$> new <*> old)
+
+-- | Computes the difference between two 'Pending' sets.
+differencePending :: Pending -> Pending -> Pending
+differencePending (Pending new) (Pending old) =
+    Pending (M.difference <$> new <*> old)
+
+-- | Adds a single element to the 'Pending' set.
+addOneToPending :: Core.TxId -> Core.TxAux -> Pending -> Pending
+addOneToPending txId aux (Pending p) = Pending (M.insert txId aux <$> p)
+
+-- Ready-to-use 'ResubmissionFunction's.
+
+-- | A 'RetryPolicy' is simply a function which instruct the 'Scheduler' when
+-- to attempt resubmitting the given 'Scheduled' item.
+type RetryPolicy = Slot -> Slot
+
+--
+-- Stock retry policies inspired by the 'retry' package.
+--
+
+constantRetry :: RetryPolicy
+constantRetry = succ
+
+exponentialBackoff :: RetryPolicy
+exponentialBackoff retries =
+    let normalised = (fromIntegral $ retries + 1) * (10.0 :: Double)
+    in retries + floor (logBase (2.0 :: Double) normalised)
+
+fibonacciBackoff :: RetryPolicy
+fibonacciBackoff base = fib (base + 1) (0, base)
+    where
+      fib 0 (a, _)    = a
+      fib !m (!a, !b) = fib (m-1) (b, a + b)
+
+-- | A very customisable resubmitter which can be configured with different
+-- retry policies.
+defaultResubmitFunction :: forall m. Monad m
+                        => (Core.TxAux -> m Bool)
+                        -> RetryPolicy
+                        -> ResubmissionFunction m
+defaultResubmitFunction send retryPolicy currentSlot scheduled oldScheduler = do
+    foldlM updateFn (emptyPending, oldScheduler) scheduled
+    where
+        updateFn :: (Evicted, Scheduler) -> Scheduled -> m (Evicted, Scheduler)
+        updateFn (evicted, acc@(Scheduler s)) entry = do
+            let Scheduled (txId, txAux, (Schedule retries)) = entry
+                rescheduled = incSubmissionCount entry
+                s' = IntMap.insertWith mappend
+                                       (retryPolicy currentSlot)
+                                       [rescheduled]
+                                       s
+            case retries >= maxRetries of
+                 True -> return (addOneToPending txId txAux evicted, acc)
+                 False -> do
+                     result <- send txAux
+                     case result of
+                         False -> return (evicted, Scheduler s')
+                         True  -> return (addOneToPending txId txAux evicted, acc)
+
+-- | A very lenient resubmitter which blindly schedules each pending transaction
+-- to be resubmitted the next slot.
```

Comment:
Slightly worried that this will end up being used in the real code base. Maybe we should just parameterize this with the retry policy also, so that we force callers to make a decision, rather than just using the most convenient function.